src/components/CommandPalette.tsx

```diff
@@ -454,42 +454,36 @@ const CommandPalette: FC = () => {
               ...(!isTouch ? { maxHeight: 'calc(100vh - 8em)', overflow: 'auto' } : null),
             })}
           >
-            {shortcuts.length > 0 ? (
-              <>
-                {(() => {
-                  const helpCommand = commandById('help')
-                  // Check if current gesture matches any command's gesture once
-                  const hasMatchingCommand = shortcuts.some(cmd => gestureInProgress === cmd.gesture)
-
-                  return shortcuts.map(shortcut => {
-                    const isHelpMatch =
-                      shortcut.id === 'help' &&
-                      (gestureInProgress as string)?.toString().endsWith(gestureString(helpCommand))
-
-                    const isCancelMatch =
-                      shortcut.id === 'cancel' &&
-                      !hasMatchingCommand &&
-                      !(gestureInProgress as string)?.toString().endsWith(gestureString(helpCommand))
-
-                    return (
-                      <CommandRow
-                        search={search}
-                        gestureInProgress={gestureInProgress as string}
-                        key={shortcut.id}
-                        last={shortcut === shortcuts[shortcuts.length - 1]}
-                        onClick={onExecute}
-                        onHover={onHover}
-                        selected={
-                          !isTouch
-                            ? shortcut === selectedShortcut
-                            : isHelpMatch || gestureInProgress == shortcut.gesture || isCancelMatch
-                        }
-                        shortcut={shortcut}
-                      />
-                    )
-                  })
-                })()}
-              </>
+            {commands.length > 0 ? (
+              commands.map(command => {
+                // Check if the current gesture sequence ends with help gesture
+                const helpCommand = commandById('help')
+                const isHelpMatch =
+                  command.id === 'help' &&
+                  (gestureInProgress as string)?.toString().endsWith(gestureString(helpCommand))
+                const hasMatchingCommand = commands.some(cmd => gestureInProgress === cmd.gesture)
```

Comment:
Moving `some` into the `map` is going to make this O(n^2) running time. We'll want to keep the `some` outside the loop to ensure it stays O(n).

See: https://github.com/cybersemics/em/pull/2762#discussion_r1909226132

Reply:
I've just put the iife-style logic back, now its outside